### PR TITLE
Fix some silenced exceptions during the test suite

### LINKED ISSUE #675: Serialization turns tuples into lists
```
>>> tup = 1,2
>>> from distributed.protocol import core
>>> core.loads(core.dumps(tup))
[1, 2]
```

It produces ignored exceptions in tests such as `distributed/tests/test_scheduler.py::test_remove_worker_from_scheduler`:

```
ERROR:tornado.application:Future <tornado.concurrent.Future object at 0x7f707583ef98> exception was never retrieved: Traceback (most recent call last):
  File "/home/antoine/tornado/tornado/gen.py", line 304, in wrapper
    yielded = next(result)
  File "/home/antoine/distributed/distributed/worker.py", line 540, in compute_one
    self.active.add(key)
TypeError: unhashable type: 'list'
```

### PR DESCRIPTION
**Pull request by pitrou**
Fixes #675 and #676.

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -758,12 +758,13 @@ def remove_worker(self, stream=None, address=None, safe=False):
             del self.worker_bytes[address]
 
             for key in self.has_what.pop(address):
-                self.who_has[key].remove(address)
-                if not self.who_has[key]:
-                    if key in self.tasks:
-                        recommendations[key] = 'released'
-                    else:
-                        recommendations[key] = 'forgotten'
+                if key in self.who_has:
```
Comment (mrocklin) [inline distributed/scheduler.py]: Why was this change necessary?  These two dictionaries should always be synchronized.

Comment (pitrou) [inline distributed/scheduler.py]: In `transition_memory_released`, there is the following:
```
            workers = self.who_has.pop(key)
            for w in workers:
                if w in self.worker_info:  # in case worker has died
                    self.has_what[w].remove(key)
```

I was seeing an ignored KeyError before the change.

Comment (mrocklin) [inline distributed/scheduler.py]: I would prefer to hunt down why these two were not matched.  I can try to find this sometime tomorrow.  What test was triggering this failure?  Generally I prefer to remain very strict in the scheduler.

Comment (pitrou) [inline distributed/scheduler.py]: `distributed/tests/test_client.py::test_scatter_hash` and `distributed/tests/test_client.py::test__scatter_types`.

Comment (pitrou) [inline distributed/scheduler.py]: I've removed the workaround.
